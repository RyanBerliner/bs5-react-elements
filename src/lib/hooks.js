import {useEffect, useRef} from 'react';

/**
 * Hook to initialize a component, bind events, and expose bootstrap apis.
 *
 * @param {Object} Component Bootstrap component object (Modal, Tooltip, etc)
 * @param {Object} config (optional) Configuration object of the component
 * @param {Object} componentRef (optional) A ref to reference the component
 * @param {Object} domRef A ref pointing to the dom element of the component
 * @param {Map} events Mapping of bootstrap event names and function handlers
 */
export function useBootstrap(
    Component,
    config,
    componentRef,
    domRef,
    events,
) {
  const initialConfig = useRef(config);

  useEffect(() => {
    const component = Component.getInstance(domRef.current) ||
      new Component(domRef.current, initialConfig.current);
    if (componentRef) {
      componentRef.current = component;
    }

    return () => component.dispose();
  }, [Component, componentRef, domRef]);

  useEffect(() => {
    const el = domRef.current;
    events.forEach((value, key) => {
      el.addEventListener(key, value);
    });

    return () => {
      events.forEach((value, key) => {
        el.removeEventListener(key, value);
      });
    };
  }, [domRef, events]);
}
