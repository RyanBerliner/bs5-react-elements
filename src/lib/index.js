export {Alert} from './Alert.jsx';
export {Carousel} from './Carousel.jsx';
export {Dropdown} from './Dropdown.jsx';
export {Modal} from './Modal.jsx';
export {Offcanvas} from './Offcanvas.jsx';
export {Popover} from './Popover.jsx';
export {Tab} from './Tab.jsx';
export {Toast} from './Toast.jsx';
export {Tooltip} from './Tooltip.jsx';
