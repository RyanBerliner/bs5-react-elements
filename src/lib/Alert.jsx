import React, {useRef, useMemo, useCallback} from 'react';
import {unmountComponentAtNode} from 'react-dom';
import PropTypes from 'prop-types';
import {Alert} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.0/components/alerts/">Bootstrap alert component.</a>
 */
function AlertComponent({
  onClosed,
  onClose,
  container,
  component,
  children,
  ...props
}) {
  const componentElement = useRef();

  const _onClosed = useCallback((event) => {
    unmountComponentAtNode(container.current);

    if (onClosed) {
      onClosed(event);
    }
  }, [onClosed]);

  const events = useMemo(() => {
    return new Map([
      ['closed.bs.alert', _onClosed],
      ['close.bs.alert', onClose],
    ]);
  }, [_onClosed, onClose]);

  useBootstrap(Alert, undefined, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

AlertComponent.propTypes = {
  /**
   * Handler for the closed.bs.alert event
   */
  onClosed: PropTypes.func,
  /**
   * Handler for the close.bs.alert event
   */
  onClose: PropTypes.func,
  /**
   * React ref refering to the container the element is wrapped in
   */
  container: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]).isRequired(),
  /**
   * React ref that will be assigned the component instance
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Alert contents
   */
  children: PropTypes.node,
};

AlertComponent.displayName = 'Alert';

export {AlertComponent as Alert};
