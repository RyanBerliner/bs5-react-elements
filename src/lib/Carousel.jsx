import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Carousel} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.2/components/carousel/">Bootstrap carousel component.</a>
 */
function CarouselComponent({
  onSlid,
  onSlide,
  component,
  config,
  children,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['slid.bs.carousel', onSlid],
      ['slide.bs.carousel', onSlide],
    ]);
  }, [onSlid, onSlide]);

  useBootstrap(Carousel, config, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

CarouselComponent.propTypes = {
  /**
   * Handler for the slid.bs.carousel event
   * @since 0.2.0
   */
  onSlid: PropTypes.func,
  /**
   * Handler for the slide.bs.carousel event
   * @since 0.2.0
   */
  onSlide: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   * @since 0.2.0
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   * @since 0.2.0
   */
  config: PropTypes.object,
  /**
   * Dropdown contents
   * @since 0.2.0
   */
  children: PropTypes.node,
};

CarouselComponent.displayName = 'Carousel';

export {CarouselComponent as Carousel};
