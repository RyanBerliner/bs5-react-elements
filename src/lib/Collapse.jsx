import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Collapse} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.0/components/collapse/">Bootstrap collapse component.</a>
 */
function CollapseComponent({
  onShown,
  onShow,
  onHidden,
  onHide,
  component,
  config,
  children,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['shown.bs.collapse', onShown],
      ['show.bs.collapse', onShow],
      ['hidden.bs.collapse', onHidden],
      ['hide.bs.collapse', onHide],
    ]);
  }, [onShown, onShow, onHidden, onHide]);

  useBootstrap(Collapse, config, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

CollapseComponent.propTypes = {
  /**
   * Handler for the shown.bs.collapse event
   */
  onShown: PropTypes.func,
  /**
   * Handler for the show.bs.collapse event
   */
  onShow: PropTypes.func,
  /**
   * Handler for the hidden.bs.collapse event
   */
  onHidden: PropTypes.func,
  /**
   * Handler for the hide.bs.collapse event
   */
  onHide: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   */
  config: PropTypes.object,
  /**
   * Collapse contents
   */
  children: PropTypes.node,
};

CollapseComponent.displayName = 'Collapse';

export {CollapseComponent as Collapse};
