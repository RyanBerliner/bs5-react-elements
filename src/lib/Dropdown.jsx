import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Dropdown} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.3/components/dropdowns/">Bootstrap dropdown component.</a>
 */
function DropdownComponent({
  onShown,
  onShow,
  onHidden,
  onHide,
  component,
  config,
  children,
  as: ElementType,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['shown.bs.dropdown', onShown],
      ['show.bs.dropdown', onShow],
      ['hidden.bs.dropdown', onHidden],
      ['hide.bs.dropdown', onHide],
    ]);
  }, [onShown, onShow, onHidden, onHide]);

  useBootstrap(Dropdown, config, component, componentElement, events);

  return (
    <ElementType ref={componentElement} {...props}>
      {children}
    </ElementType>
  );
}

DropdownComponent.propTypes = {
  /**
   * Handler for the shown.bs.dropdown event
   * @since 0.2.0
   */
  onShown: PropTypes.func,
  /**
   * Handler for the show.bs.dropdown event
   * @since 0.2.0
   */
  onShow: PropTypes.func,
  /**
   * Handler for the hidden.bs.dropdown event
   * @since 0.2.0
   */
  onHidden: PropTypes.func,
  /**
   * Handler for the hide.bs.dropdown event
   * @since 0.2.0
   */
  onHide: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   * @since 0.2.0
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   * @since 0.2.0
   */
  config: PropTypes.object,
  /**
   * Dropdown contents
   * @since 0.2.0
   */
  children: PropTypes.node,
  /**
   * The element type of the dropdown
   * @since 0.2.0
   */
  as: PropTypes.elementType,
};

DropdownComponent.defaultProps = {
  as: 'button',
};

DropdownComponent.displayName = 'Dropdown';

export {DropdownComponent as Dropdown};
