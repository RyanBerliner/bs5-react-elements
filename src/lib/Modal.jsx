import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Modal} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.3/components/modal/">Bootstrap modal component.</a>
 */
function ModalComponent({
  onShown,
  onShow,
  onHidden,
  onHide,
  onHidePrevented,
  component,
  config,
  children,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['shown.bs.modal', onShown],
      ['show.bs.modal', onShow],
      ['hidden.bs.modal', onHidden],
      ['hide.bs.modal', onHide],
      ['hidePrevented.bs.modal', onHidePrevented],
    ]);
  }, [onShown, onShow, onHidden, onHide, onHidePrevented]);

  useBootstrap(Modal, config, component, componentElement, events);

  return (
    <div ref={componentElement} {...props}>
      {children}
    </div>
  );
}

ModalComponent.propTypes = {
  /**
   * Handler for the shown.bs.modal event
   * @since 0.1.0
   */
  onShown: PropTypes.func,
  /**
   * Handler for the show.bs.modal event
   * @since 0.1.0
   */
  onShow: PropTypes.func,
  /**
   * Handler for the hidden.bs.modal event
   * @since 0.1.0
   */
  onHidden: PropTypes.func,
  /**
   * Handler for the hide.bs.modal event
   * @since 0.1.0
   */
  onHide: PropTypes.func,
  /**
   * Handler for the hidePrevented.bs.modal event
   * @since 0.1.0
   */
  onHidePrevented: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   * @since 0.1.0
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Component configuration
   * @since 0.1.0
   */
  config: PropTypes.object,
  /**
   * Modal contents
   * @since 0.1.0
   */
  children: PropTypes.node,
};

ModalComponent.displayName = 'Modal';

export {ModalComponent as Modal};
