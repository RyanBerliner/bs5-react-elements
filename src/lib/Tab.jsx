import React, {useRef, useMemo} from 'react';
import PropTypes from 'prop-types';
import {Tab} from 'bootstrap';
import {useBootstrap} from './hooks.js';

/**
 * Wrapper for the <a href="https://getbootstrap.com/docs/5.0/components/navs-tabs/">Bootstrap tab component.</a>
 */
function TabComponent({
  onShown,
  onShow,
  onHidden,
  onHide,
  component,
  children,
  as: ElementType,
  ...props
}) {
  const componentElement = useRef();

  const events = useMemo(() => {
    return new Map([
      ['shown.bs.tab', onShown],
      ['show.bs.tab', onShow],
      ['hidden.bs.tab', onHidden],
      ['hide.bs.tab', onHide],
    ]);
  }, [onShown, onShow, onHidden, onHide]);

  useBootstrap(Tab, undefined, component, componentElement, events);

  return (
    <ElementType ref={componentElement} {...props}>
      {children}
    </ElementType>
  );
}

TabComponent.propTypes = {
  /**
   * Handler for the shown.bs.tab event
   * @since 0.2.0
   */
  onShown: PropTypes.func,
  /**
   * Handler for the show.bs.tab event
   * @since 0.2.0
   */
  onShow: PropTypes.func,
  /**
   * Handler for the hidden.bs.tab event
   * @since 0.2.0
   */
  onHidden: PropTypes.func,
  /**
   * Handler for the hide.bs.tab event
   * @since 0.2.0
   */
  onHide: PropTypes.func,
  /**
   * React ref that will be assigned the component instance
   * @since 0.2.0
   */
  component: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({current: PropTypes.any}),
  ]),
  /**
   * Tab contents
   * @since 0.2.0
   */
  children: PropTypes.node,
  /**
   * The element type of the tab
   * @since 0.2.0
   */
  as: PropTypes.elementType,
};

TabComponent.defaultProps = {
  as: 'button',
};

TabComponent.displayName = 'Tab';

export {TabComponent as Tab};
