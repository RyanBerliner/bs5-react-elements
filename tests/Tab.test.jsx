import React, {useEffect, useRef, forwardRef} from 'react';
import {Tab} from '../src/lib/Tab';
import {render, screen, fireEvent} from '@testing-library/react';

/**
 * A boilerplate tabs widget, with 2 tabs.
 * Only spreads props onto the second tab.
 */
function TestTabs(props) {
  return <>
    <ul className="nav nav-tabs" role="tablist">
      <li className="nav-item" role="presentation">
        <Tab
          className="nav-link active"
          id="tab-1"
          data-bs-toggle="tab"
          data-bs-target="#panel-1"
          type="button"
          role="tab"
          aria-controls="panel-1"
          aria-selected="true">Tab 1</Tab>
      </li>
      <li className="nav-item" role="presentation">
        <Tab
          className="nav-link"
          id="tab-2"
          data-bs-toggle="tab"
          data-bs-target="#panel-2"
          type="button"
          role="tab"
          aria-controls="panel-2"
          aria-selected="false"
          {...props}>Tab 2</Tab>
      </li>
    </ul>

    <div className="tab-content">
      <div
        className="tab-pane active"
        id="panel-1"
        role="tabpanel"
        aria-labelledby="tab-1">Lorem ispum 1</div>
      <div
        className="tab-pane"
        id="panel-2"
        role="tabpanel"
        aria-labelledby="tab-2">Lorem ipsum 2</div>
    </div>
  </>;
}

/**
 * Clicks a specific tab
 *
 * @param {integer} number The number of the tab to click
 */
function clickTab(number) {
  fireEvent.click(screen.getByText(`Tab ${number}`));
}

// Timeouts are used to prevent disposing of bootstrap components that are in
// the middle of a transition. This is a bootstrap bug. TODO: Submit an issue.

describe('events bound to props', () => {
  test('show.bs.tab event bound to onShow prop', (done) => {
    /**
     * show.bs.tab event handler.
     *
     * @param {Object} event
     */
    function onShow(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    render(<TestTabs onShow={onShow} />);
    clickTab(2);
  });

  test('shown.bs.tab event bound to onShown prop', (done) => {
    /**
     * shown.bs.tab event handler.
     *
     * @param {Object} event
     */
    function onShown(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    render(<TestTabs onShown={onShown} />);
    clickTab(2);
  });

  test('hide.bs.tab event bound to onHide prop', (done) => {
    /**
     * hide.bs.tab event handler.
     *
     * @param {Object} event
     */
    function onHide(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    /**
     * shown.bs.tab event handler
     */
    function onShown() {
      clickTab(1);
    }

    render(<TestTabs onShown={onShown} onHide={onHide} />);
    clickTab(2);
  });

  test('hidden.bs.tab event bound to onHidden prop', (done) => {
    /**
     * hidden.bs.tab event handler.
     *
     * @param {Object} event
     */
    function onHidden(event) {
      expect(event).toBeTruthy();
      setTimeout(() => done(), 500);
    }

    /**
     * shown.bs.tab event handler
     */
    function onShown() {
      clickTab(1);
    }

    render(<TestTabs onShown={onShown} onHidden={onHidden} />);
    clickTab(2);
  });
});

describe('bootstrap apis', () => {
  test('component object exposed', (done) => {
    /**
     * A tabs widget that will call the bootstrap show() method immediately on
     * mount.
     */
    function TestComponent(props) {
      const component = useRef();
      useEffect(() => component.current.show());
      return <TestTabs component={component} {...props} />;
    }

    /**
     * show.bs.tab event handler.
     */
    function onShow() {
      setTimeout(() => done(), 500);
    }

    render(<TestComponent onShow={onShow} />);
  });
});

describe('markup', () => {
  test('minimum markup renders as expected', () => {
    const {container} = render(<Tab />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('places children in correct element', () => {
    const {container} = render(<Tab>child</Tab>);
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('support additional attributes, place them correctly', () => {
    const {container} = render(<Tab data-lorem="ipsum" />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('supports custom tagname by string', () => {
    const {container} = render(<Tab as="a" />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('support custom tagname with components supporting ref forwards', () => {
    /**
     * A component that renders a link, and importantly supports ref forwarding.
     */
    const Link = forwardRef((props, ref) => {
      const {children, ...rest} = props;
      return <a ref={ref} {...rest}>{children}</a>;
    });

    Link.displayName = 'Link';

    Link.propTypes = {
      children: PropTypes.node,
    };

    const {container} = render(<Tab as={Link}>child</Tab>);
    expect(container.innerHTML).toMatchSnapshot();
  });
});
