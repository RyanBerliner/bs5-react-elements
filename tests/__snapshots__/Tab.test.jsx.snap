// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markup minimum markup renders as expected 1`] = `"<button></button>"`;

exports[`markup places children in correct element 1`] = `"<button>child</button>"`;

exports[`markup support additional attributes, place them correctly 1`] = `"<button data-lorem=\\"ipsum\\"></button>"`;

exports[`markup support custom tagname with components supporting ref forwards 1`] = `"<a>child</a>"`;

exports[`markup supports custom tagname by string 1`] = `"<a></a>"`;
